### PR TITLE
Encoding layer

### PR DESCRIPTION
**Pull request by veewee**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | yes
| Fixed issues | 

#### Summary

Since the request and response types are very flexible, it seems like a better idea to provide composable encoders / decoders that can transform the various request/response types into actual PSR-7 Request / Response body payloads.

**This PR introduces a BC break:**

`JsonTransport` and `AsyncJsonTransport` are no longer available and are replaced by `AsyncEncodedTransport` and `EncodedTransport`. However, we do provide a pre-configured json transport configuration which is compatible with the previously available one. Please use one of the following instead:

- `\Phpro\HttpTools\Transport\Presets\JsonPreset::sync()`
- `\Phpro\HttpTools\Transport\Presets\JsonPreset::async()`

### CODE REVIEW

File: src/Encoding/Json/JsonEncoder.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * @implements EncoderInterface<array|null>
+ */
+final class JsonEncoder implements EncoderInterface
+{
+    private StreamFactoryInterface $streamFactory;
+
+    public function __construct(StreamFactoryInterface $streamFactory)
+    {
+
```
Comment (pottink) [inline src/Encoding/Json/JsonEncoder.php:22]: obsolete newline

File: src/Encoding/Stream/StreamEncoder.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use Phpro\HttpTools\Encoding\EncoderInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * @implements EncoderInterface<StreamInterface>
+ */
+final class StreamEncoder implements EncoderInterface
+{
+    public static function createWithAutodiscoveredPsrFactories(): self {
```
Comment (pottink) [inline src/Encoding/Stream/StreamEncoder.php:16]: { on newline

**Review by pottink — APPROVED** [review]
Very nice improvements @veewee ! Good job!

**Comment by veewee** [comment]
Thanks for the review. Looks like the cs-fixer config is invalid.
Will perform those CS related changes all at once in another PR (#10) that fixes the config.

**Pull request merged**